### PR TITLE
feat: test in Firefox and headless mode

### PR DESCRIPTION
**Pull request by ValeraS**
#### Pre-Submission Checklist

- [x] Your pull request targets the `master` branch of freeCodeCamp/testable-projects-fcc
- [ ] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] Your changes have been tested either locally or using a newly created CDN based on your fork's testable-projects-fcc/build/bundle.js file
- [ ] Working changes have been added to the build by running `npm run build`

#### Type of Change

- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.

- [ ] Closes currently open issue: Closes #XXX

#### Description

- automated testing in Firefox
- headless mode
- muted sounds

### CODE REVIEW

**Comment by ValeraS** [comment]
@tchaffee @Thibaut I'm interested in your opinion.

CONTRIBUTING.md is not yet corrected.

**Comment by tchaffee** [comment]
@ValeraS Thanks for the PR. We are interested in supporting Firefox eventually. In terms of timing I would like to get the issues close to zero before we throw another browser in the mix. I'm curious though, do all the tests also pass on Firefox?

**Comment by ValeraS** [comment]
@tchaffee Yes, all tests pass in Firefox. You can see it in Travis CI.

File: .travis.yml [hunk 1]
```diff
@@ -4,14 +4,10 @@ node_js:
...
   chrome: stable
+  firefox: latest
 before_install:
   - if [[ `npm -v` != 3* ]]; then npm i -g npm@3; fi
-  - "/sbin/start-stop-daemon --start --quiet --pidfile /tmp/custom_xvfb_99.pid --make-pidfile --background --exec /usr/bin/Xvfb -- :99 -ac -screen 0 1920x1080x16"
-  # give xvfb some time to start
-  - sleep 3
 before_script:
-  # Same as what we specified for Xvfb.
-  - "export DISPLAY=:99.0"
```
Comment (tbushman) [inline .travis.yml]: Is this needed for screenshots?

Comment (ValeraS) [reply .travis.yml]: It was needed for tests run on the virtual display without popping up the browser.
But now headless mode is used and it's not needed.

**Comment by tchaffee** [comment]
@ValeraS This is a good enhancement but at the moment I am very busy with launching my startup so I hesitate to merge this and make any announcements about it. 

Would you be willing to resolve any new issues that come up related to Firefox for the next couple of months? If you can commit to that, I would feel more comfortable merging this. 

We should also update the forkable CodePen to mention we now support FireFox. I can do that. And Quincy might want to announce it. It's a major goal and kind of big news.

Let me know your thoughts.

**Comment by ValeraS** [comment]
@tchaffee I'll do anything in my power to help.

- Pens don't need to be updated. But we must decide what to do with the alert. Are we going to show it in other browsers or delete it? Maybe we need to show it only in IE, as CodePen does?
- Contributing.md needs to be changed. Do we delete information related to Xvfb (Running Chrome in a headless environment and Capturing video of the tests)?

What to do with other PRs? Can anyone check them out?

**Comment by tchaffee** [comment]
>  I'll do anything in my power to help.

I'll take that as a commitment then. In which case, welcome to the regular team. Your contributions have already improved things and I'm looking forward to working with you on a continued basis for the near future.

> Pens don't need to be updated

This is in the CodePen they fork. It should be updated to include FireFox:

```
<!-- 

Hello Camper!

For now, the test suite only works in Chrome! Please read the README below in the JS Editor before beginning. Feel free to delete this message once you have read it. Good luck and Happy Coding! 

- The freeCodeCamp Team 

-->
```

> But we must decide what to do with the alert. Are we going to show it in other browsers or delete it? Maybe we need to show it only in IE, as CodePen does?

It should be shown in every browser except Chrome and FireFox since those will be the only two we support for now.

> Contributing.md needs to be changed. Do we delete information related to Xvfb (Running Chrome in a headless environment and Capturing video of the tests)?

That raises a couple of questions. 

1. In addition to using Xvfb, I sometimes intentionally do not use it so I can watch the automated tests when there is a bug. I can often see the problem just by watching. Is that going to change? I'll no longer be able to watch the automated tests? That might make debugging harder. Let's discuss.

1. Can we still capture video? 

1. If not video, then why do screenshots still work? I remember seeing the code for screenshots still in this PR, so I'm assuming we can capture screenshots and you've tested it.

> What to do with other PRs? Can anyone check them out?

I should be able to work on this project a little each day so we will definitely make progress. Please be patient and we will get there pretty quickly. I'm focusing on this PR for now because I started on it already, and it's a big step forward for the project, accomplishing a major goal we've had for a long time.

**Comment by ValeraS** [comment]
1. I propose to add an option for `headless` into `setup.js`.
2. I need time to think about it. With the first option, we can capture a video as before.
3. A screenshot is done by the browser.

**Comment by tchaffee** [comment]
> I propose to add an option for headless into setup.js.

That would work for me.

> I need time to think about it. With the first option, we can capture a video as before.

That's fine then. At the moment we don't capture video from failed Travis builds, but that's something to keep in mind as a future goal.

> A screenshot is done by the browser.

Good enough.

Except for the Travis video which isn't an immediate need (just an idea) I think if you put the headless option in `setup.js` that will be good enough?

**Comment by ValeraS** [comment]
@tchaffee Could you look at contributing.md and help fix it?

**Comment by tchaffee** [comment]
@ValeraS Sure, but I don't know what needs fixing? What areas do you have concerns about?

**Comment by ValeraS** [comment]
@tchaffee In common, I've corrected somehow but stuck. Could you read it and give me suggestions for enhancement?

**Comment by tchaffee** [comment]
Ok, I will. Just give me some time. The next couple of days are very busy. But I will find time.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -151,29 +151,11 @@ Leave that running. And in a new window, you can run the tests:
...
-1. The "Tests" button is clicked to see the test output.
-1. A screenshot of the results is saved.
-1. Move on to the next CodePen by returning to step 1.
-
 You will find screenshots of all the test results in the `test/screenshots`
 directory (or the directory you configured in your `test/setup.js` file).
 
 #### Testing tips
 
-**Turn your volume down**
```
Comment (tchaffee) [inline CONTRIBUTING.md]: Why did this get removed? Is there no sound when in headless mode?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -186,19 +168,34 @@ directory no matter where you are in the project directory:
...
 
-For the names used by the grep option, see the `name` attribute in the
-`tests` array at the top of the `tests/codepen-project.js` file.
+For the names used by the grep option, see the `name` attribute in the default
+export object of the `src/project-tests/index.js` file.
+
+**Running tests without headless mod**
 
-**Running Chrome in a headless environment**
+By default tests run in the headless mod. If you need to see how it works on
```
Comment (tchaffee) [inline CONTRIBUTING.md]: Should be "headless mode".

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -186,19 +168,34 @@ directory no matter where you are in the project directory:
...
+your screen, you need to disable the headless mod. For this run the tests like
```
Comment (tchaffee) [inline CONTRIBUTING.md]: Should be "headless mode".

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -186,19 +168,34 @@ directory no matter where you are in the project directory:
...
+1. The "Tests" button is clicked to see the test output.
+1. A screenshot of the results is saved.
+1. Move on to the next CodePen by returning to step 1.
 
-We tried to get headless Chrome (very new as of July 2017) to work but it had a
-lot of problems.
+**Running tests on a virtual screen**
 
-In the meantime, you can use the tried and trusted method of running a virtual
+You can use the tried and trusted method of running a virtual
```
Comment (tchaffee) [inline CONTRIBUTING.md]: I believe this is an alternative to headless mode? If so, we should let people know it's optional. Something like "If for some reason headless mode doesn't work for you, you can use the tried and trusted method of running a virtual..."

File: CONTRIBUTING.md (hunk 2 above)
Comment (ValeraS) [reply CONTRIBUTING.md]: No, it is not because of headless mode. I turn off sounds in settings for Chrome and Firefox.

Comment (tchaffee) [reply CONTRIBUTING.md]: I'm not sure what you mean by this? Do you personally turn off sounds on your preferences? Or is there some code here that turns it off so no one will ever hear the sounds during testing?

Comment (ValeraS) [reply CONTRIBUTING.md]: some code

Comment (tchaffee) [reply CONTRIBUTING.md]: Cool. Can you point me to that code? I couldn't find it.

Comment (tchaffee) [reply CONTRIBUTING.md]: Nevermind, I found it. It was hidden because of github not showing large diffs.

File: test/automate/automate-utils.js [hunk 6]
```diff
@@ -14,285 +14,292 @@
...
 
   // If all of the project tests pass, this is set to true.
   let success = false;
 
-  const chrome = require('selenium-webdriver/chrome');
+  // headless mode for chrome and firefox is enabled by default
+  let headless = true;
+  if (typeof process.env.HEADLESS !== 'undefined') {
+    // eslint-disable-next-line eqeqeq
+    headless = process.env.HEADLESS == true;
```
Comment (tchaffee) [inline test/automate/automate-utils.js]: Can you please use triple equal (===) here? I'm suprised ESLint didn't warn you about this.

File: test/automate/automate-utils.js [hunk 7]
```diff
@@ -14,285 +14,292 @@
...
+  }
+
+  let capabilities;
+
+  if (browser === 'chrome') {
+
+    // Set up Chrome options.
```
Comment (tchaffee) [inline test/automate/automate-utils.js]: Can you document what the options are and why we are setting them? Why no-mute, why log level 3? etc.

File: test/automate/automate-utils.js [hunk 8]
```diff
@@ -14,285 +14,292 @@
...
+      firefoxOptions.headless();
+    }
+
+    capabilities = firefoxOptions.toCapabilities();
+
+  } else {
+    throw new Error(`Browser ${browser} is not supported`);
+  }
+
+  capabilities.set('acceptInsecureCerts', true);
```
Comment (tchaffee) [inline test/automate/automate-utils.js]: Why are we setting this? Could you document what happens if we don't have this line?

File: test/automate/automate-utils.js [hunk 9]
```diff
@@ -14,285 +14,292 @@
...
+          );
+        }
+        return null;
+      });
+  };
+
   // Locates an element and then clicks it.
-  const clickElement = function(locator) {
-    return driver.wait(
+  const clickElement = locator => driver.wait(
```
Comment (tchaffee) [inline test/automate/automate-utils.js]: This is a code improvement and isn't related to this PR. In the future you should put code improvements in a separate PR. If we have to reverse this PR we would lose the code improvement with it. I'm not going to let it stop us from merging this PR, but please put in the effort in the future to keep PRs very focused.

File: test/automate/automate-utils.js [hunk 10]
```diff
@@ -14,285 +14,292 @@
...
-      saveScreenshot(driver, screenshotDir, name, `ERROR-${error.name}`);
-    } else {
-      saveScreenshot(driver, screenshotDir, name, `ERROR-${error.name}`);
+  const errorFunc = error => {
       console.error(error);
-    }
+      return saveScreenshot(name, `ERROR-${error.name}`);
+  };
+
+  // Alert appears only once.
```
Comment (tchaffee) [inline test/automate/automate-utils.js]: This looks like another improvement? I think we were having problems in the past with the tests failing due to alerts? Good improvement, but like above, please try to put improvements in a separate PR in the future.

File: test/automate/automate-utils.js [hunk 11]
```diff
@@ -14,285 +14,292 @@
...
+  return driver.manage().window().setPosition(0, 0)
+  .then(() => driver.manage().window()
+    .setSize(browserMaxWidth, browserMaxHeight)
+  )
 
   // Get the specified URL.
-  driver.get(URL);
+  .then(() => driver.get(URL))
+
+  // Will be not needed after deploy
```
Comment (tchaffee) [inline test/automate/automate-utils.js]: What does this mean "will not be needed after deploy"?

**Comment by tchaffee** [comment]
@ValeraS any chance you'll have time soon to look at the requested changes so we can get this PR merged?

**Comment by ValeraS** [comment]
@tchaffee Excuse me for the delay. I have a problem with one test in firefox when running on Travis CI. And I do not understand why. When the test runs locally both on Windows and Ubuntu everything works fine, but not on Travis CI.

We can `pending` this test. What do you think?

**Comment by tchaffee** [comment]
Which test is it? I can try to help figure it out and fix it.

**Comment by ValeraS** [comment]
image

**Comment by tchaffee** [comment]
Might have to do with running it headless and turning off audio. I will take a look.

**Comment by ValeraS** [comment]
you can check what I have tried in this branch https://github.com/ValeraS/testable-projects-fcc/commits/test/firefox2

File: test/automate/automate-utils.js (hunk 11 above)
Comment (ValeraS) [reply test/automate/automate-utils.js]: For now, in the headless mode, an alert is displayed for both Firefox and Chrome. After deployment, the alert will not be displayed. So this check will be not needed

File: test/automate/automate-utils.js (hunk 10 above)
Comment (ValeraS) [reply test/automate/automate-utils.js]: Before deployment of this PR, tests don't pass without this.

**Comment by ValeraS** [comment]
@tchaffee It's done.

File: test/automate/automate-utils.js (hunk 7 above)
Comment (ValeraS) [reply test/automate/automate-utils.js]: documented

File: test/automate/automate-utils.js (hunk 8 above)
Comment (ValeraS) [reply test/automate/automate-utils.js]: documented

**Comment by tchaffee** [comment]
@ValeraS Thanks for all your work on this. It's a big PR, and a big move forward for the project.

**Comment by tchaffee** [comment]
@raisedadead @QuincyLarson With this PR from @ValeraS we now officially support the FireFox browser for our testable projects. If you feel an announcement is appropriate then please go ahead with that. @ValeraS deserves all the credit for this.

**Pull request merged**